[SERD-1445] return JSON from run template

arthir opened the pull request:
Modified run_template() so that if specified, the JSON values are returned instead of the associated file ids

arthir commented on R/scripts.R:
```diff
@@ -99,12 +102,24 @@ run_civis <- function(expr, ...) {
   await(scripts_get_custom_runs, id = job$id, run_id = run$id)
-  fetch_output_file_ids(civis_script(job$id, run$id))
+  if (JSONValue) {
+     output <- fetch_output(civis_script(job$id, run$id))
```
This already seems to have the JSON values so decided to use this instead of calling json_values_get()

arthir commented on R/scripts.R:
```diff
@@ -99,12 +102,24 @@ run_civis <- function(expr, ...) {
+  if (JSONValue) {
+     output <- fetch_output(civis_script(job$id, run$id))
+     names <- lapply(output, function(o) o$name)
+     file_values <- stats::setNames(lapply(output, function(o) o$value), names)
```
Also this means that the output should mimic the output of the file_ids look like a JSON with
{
"output": ....,
"output2": .....
}

arthir replied:
Changed so only single JSON output is returned

arthir commented on R/scripts.R:
```diff
@@ -99,12 +106,31 @@ run_civis <- function(expr, ...) {
+        cat("Error in returning JSON outputs of template run -- no JSON output")
+	return()
+     }
+     return(json_output[[1]]$value)
```
Returning something if and only if there is exactly 1 JSON output

keithing replied:
My instinct here (if `JSONValue=`TRUE`) would be the following:

* If there is 1 JSONValue return that value
* If there is more than 1 JSONValue, return the first and warn that only the first is returned.
* If there are no JSONValues returned, warn that no value was returned and return `list()`

arthir replied:
That makes sense as well. I wasn't positive what the best behavior was for case 2. 
In the latest commit, I changed it to the behavior described.

keithing commented on R/scripts.R:
```diff
@@ -99,12 +106,31 @@ run_civis <- function(expr, ...) {
+     output <- fetch_output(civis_script(job$id, run$id))
+     json_output <- output[sapply(output, function(o) o$objectType=="JSONValue")]
+     if (length(json_output) > 1) {
+        cat("Error in returning JSON outputs of template run -- too many JSON outputs")
```
`warning` seems more appropriate than `cat` here.

arthir replied:
👍 changed to warning.

arthir commented:
@keithing Fixed this and ready for re-review

keithing commented:
(cc @patr1ckm, we're extending your `run_template` function)

keithing commented on R/scripts.R:
```diff
@@ -88,8 +88,15 @@ run_civis <- function(expr, ...) {
+#'         If JSONValue is TRUE, JSON values of first JSON run output is returned.
+#'           If there are no JSON outputs, nothing is returned
+#'           If there are more than 1 JSON outputs, error message is printed 
+#'             and nothing is returned.
```
This is stale since your update to the behavior for > 1 JSON output

arthir replied:
and.. i think we're there now.

keithing commented on R/scripts.R:
```diff
@@ -99,12 +107,33 @@ run_civis <- function(expr, ...) {
 #' run_template(id, arguments = list(arg1 = 1, arg2 = 2), ...)
+#'
+#' # Run the template and return JSON value outputs
+#' run_template(id, arguments = list(arg1 = 1, arg2 = 2), JSONValue=TRUE...)
```
missed a comma here:

`JSONValue=TRUE, ...`

keithing replied:
Also, for consistency this should be `JSONValue = TRUE`

arthir replied:
fixed it.

arthir commented:
> Code looks good to me. Like in the python client, could you also add a test and update the Changelog here as well?

Added tests and updated changelog
